storage: TestMergeQueue failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/3e1c440300e4b41858e70ec9e44663bf35ec2134

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestMergeQueue PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=963846&tab=buildLog

```
I181013 06:34:51.311348 48860 storage/replica_command.go:300  [s1,r3/1:{b-/Max}] initiating a split of this range at key "c" [r4]
--- FAIL: TestMergeQueue (7.74s)
=== RUN   TestMergeQueue/combined-threshold
I181013 06:34:57.209404 49382 storage/replica_command.go:300  [s1,r2/1:{a-c}] initiating a split of this range at key "b" [r9]
==================
WARNING: DATA RACE
Write at 0x00c4204549a0 by goroutine 372:
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue.func12()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2558 +0x2f4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4204549a0 by goroutine 704:
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleReplicatedEvalResult()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:6885 +0x23e
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleEvalResultRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:823 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5638 +0xa2c
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4463 +0x17f3
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4151 +0xeb
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3768 +0x103
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:214 +0x343
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14a

Goroutine 372 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2545 +0x1450
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 704 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:164 +0x149
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3844 +0xcd
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1358 +0xcaf
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).addStore()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:790 +0xce7
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:296 +0xd0e
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2449 +0x2e2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
I181013 06:34:57.274289 49387 storage/replica_command.go:432  [merge,s1,r2/1:{a-b}] initiating a merge of r9:{b-c} [(n1,s1):1, next=2, gen=0] into this range
I181013 06:34:57.304737 49002 storage/store.go:2594  [merge,s1,r2/1:{a-b},txn=a81aadec] removing replica r9/1
    --- FAIL: TestMergeQueue/combined-threshold (1.08s)
    	testing.go:730: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/396077240ff8bbae6c4327ae78e66d6a4e789d2a

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestMergeQueue PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=965113&tab=buildLog

```
--- FAIL: TestMergeQueue (7.94s)
=== RUN   TestMergeQueue/combined-threshold
I181014 08:00:34.526561 49258 storage/replica_command.go:300  [s1,r2/1:{a-c}] initiating a split of this range at key "b" [r9]
==================
WARNING: DATA RACE
Write at 0x00c4202dd7c0 by goroutine 159:
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue.func12()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2558 +0x2f4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4202dd7c0 by goroutine 776:
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleReplicatedEvalResult()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:6885 +0x23e
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleEvalResultRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:823 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5638 +0xa2c
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4463 +0x17f3
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4151 +0xeb
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3768 +0x103
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:214 +0x343
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14a

Goroutine 159 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2545 +0x1450
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 776 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:164 +0x149
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3844 +0xcd
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1358 +0xcaf
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).addStore()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:790 +0xce7
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:296 +0xd0e
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2449 +0x2e2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
I181014 08:00:34.573405 49270 storage/replica_command.go:432  [merge,s1,r2/1:{a-b}] initiating a merge of r9:{b-c} [(n1,s1):1, next=2, gen=0] into this range
I181014 08:00:34.605153 48921 storage/store.go:2594  [merge,s1,r2/1:{a-b},txn=6b8f9070] removing replica r9/1
    --- FAIL: TestMergeQueue/combined-threshold (1.15s)
    	testing.go:730: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/85ccbd67d3cb5b7d18ceade231fd01d63579bacd

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestMergeQueue PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=966255&tab=buildLog

```
I181015 06:57:38.401314 48738 storage/replica_command.go:300  [s1,r3/1:{b-/Max}] initiating a split of this range at key "c" [r4]
--- FAIL: TestMergeQueue (7.64s)
=== RUN   TestMergeQueue/combined-threshold
I181015 06:57:44.201807 49270 storage/replica_command.go:300  [s1,r2/1:{a-c}] initiating a split of this range at key "b" [r8]
==================
WARNING: DATA RACE
Write at 0x00c4200e27e0 by goroutine 222:
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue.func12()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2558 +0x2f4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4200e27e0 by goroutine 174:
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleReplicatedEvalResult()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:6885 +0x23e
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleEvalResultRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:823 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5638 +0xa2c
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4463 +0x17f3
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4151 +0xeb
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3768 +0x103
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:214 +0x343
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14a

Goroutine 222 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2545 +0x1450
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 174 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xba
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:164 +0x149
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3844 +0xcd
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1358 +0xcaf
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).addStore()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:790 +0xce7
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContext).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:296 +0xd0e
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_merge_test.go:2449 +0x2e2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
I181015 06:57:44.237234 49331 storage/replica_command.go:432  [merge,s1,r2/1:{a-b}] initiating a merge of r8:{b-c} [(n1,s1):1, next=2, gen=0] into this range
I181015 06:57:44.267851 48967 storage/store.go:2594  [merge,s1,r2/1:{a-b},txn=ef73159f] removing replica r8/1
    --- FAIL: TestMergeQueue/combined-threshold (1.09s)
    	testing.go:730: race detected during execution of test

```

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -48,6 +48,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
@@ -2555,6 +2556,7 @@ func TestMergeQueue(t *testing.T) {
 		verifyUnmerged(t)
 
 		// Once the maximum size threshold is increased, the merge can occur.
+		zone = *protoutil.Clone(&zone).(*config.ZoneConfig)
 		*zone.RangeMaxBytes += 1
 		setZones(zone)
 		store.ForceMergeScanAndProcess()

